File: backend/app/schemas/hutbe.py
from pydantic import BaseModel, Field, ConfigDict
from datetime import date, datetime
from typing import Optional
from uuid import UUID


class HutbeBase(BaseModel):
    """Base Hutbe schema."""
    title: str = Field(..., min_length=1, max_length=500)
    content: str = Field(..., min_length=1)
    summary: Optional[str] = None
    date: date
    year: int
    category: Optional[str] = Field(None, max_length=100)
    reading_time_minutes: Optional[int] = None
    source_url: Optional[str] = Field(None, max_length=500)
    is_featured: bool = False


class HutbeCreate(HutbeBase):
    """Schema for creating a Hutbe."""
    pass


class HutbeUpdate(BaseModel):
    """Schema for updating a Hutbe."""
    title: Optional[str] = Field(None, min_length=1, max_length=500)
    content: Optional[str] = Field(None, min_length=1)
    summary: Optional[str] = None
    date: Optional[date] = None
    year: Optional[int] = None
    category: Optional[str] = Field(None, max_length=100)
    reading_time_minutes: Optional[int] = None
    source_url: Optional[str] = Field(None, max_length=500)
    is_featured: Optional[bool] = None


class HutbeInDB(HutbeBase):
    """Schema for Hutbe in database."""
    id: UUID
    created_at: datetime
    updated_at: datetime
    
    model_config = ConfigDict(from_attributes=True)


class HutbeResponse(HutbeInDB):
    """Schema for Hutbe API response."""
    pass


class HutbeListItem(BaseModel):
    """Schema for Hutbe list item (minimal data)."""
    id: UUID
    title: str
    date: date
    year: int
    category: Optional[str] = None
    reading_time_minutes: Optional[int] = None
    is_featured: bool = False
    
    model_config = ConfigDict(from_attributes=True)


class HutbeSearchResult(BaseModel):
    """Schema for search results."""
    total: int
    items: list[HutbeListItem]
    page: int
    page_size: int
    total_pages: int


class YearStats(BaseModel):
    """Schema for year statistics."""
    year: int
    count: int
    categories: dict[str, int]


class CategoryStats(BaseModel):
    """Schema for category statistics."""
    category: str
    count: int

File: flutter_app/lib/models/hutbe.dart
class Hutbe {
  final String id;
  final String title;
  final String content;
  final String? summary;
  final DateTime date;
  final int year;
  final String? category;
  final int? readingTimeMinutes;
  final String? sourceUrl;
  final bool isFeatured;
  final DateTime createdAt;
  final DateTime updatedAt;

  Hutbe({
    required this.id,
    required this.title,
    required this.content,
    this.summary,
    required this.date,
    required this.year,
    this.category,
    this.readingTimeMinutes,
    this.sourceUrl,
    required this.isFeatured,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Hutbe.fromJson(Map<String, dynamic> json) {
    return Hutbe(
      id: json['id'] as String,
      title: json['title'] as String,
      content: json['content'] as String,
      summary: json['summary'] as String?,
      date: DateTime.parse(json['date'] as String),
      year: json['year'] as int,
      category: json['category'] as String?,
      readingTimeMinutes: json['reading_time_minutes'] as int?,
      sourceUrl: json['source_url'] as String?,
      isFeatured: json['is_featured'] as bool,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'content': content,
      'summary': summary,
      'date': date.toIso8601String(),
      'year': year,
      'category': category,
      'reading_time_minutes': readingTimeMinutes,
      'source_url': sourceUrl,
      'is_featured': isFeatured,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }
}

class HutbeListItem {
  final String id;
  final String title;
  final DateTime date;
  final int year;
  final String? category;
  final int? readingTimeMinutes;
  final bool isFeatured;

  HutbeListItem({
    required this.id,
    required this.title,
    required this.date,
    required this.year,
    this.category,
    this.readingTimeMinutes,
    required this.isFeatured,
  });

  factory HutbeListItem.fromJson(Map<String, dynamic> json) {
    return HutbeListItem(
      id: json['id'] as String,
      title: json['title'] as String,
      date: DateTime.parse(json['date'] as String),
      year: json['year'] as int,
      category: json['category'] as String?,
      readingTimeMinutes: json['reading_time_minutes'] as int?,
      isFeatured: json['is_featured'] as bool,
    );
  }
}
